### PR TITLE
Fix source and cell sets objects copy

### PR DESCRIPTION
**Pull request by cosa65**
# Description

# Details
#### URL to issue
N/A

#### Link to staging deployment URL (or set N/A)
https://ui-martinfosco-pipeline166.scp-staging.biomage.net/

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: pipeline-runner/R/copy-1-copy_s3_objects.R [hunk 1]
```diff
@@ -85,16 +92,19 @@ copy_filtered_cells <- function(
...
 
-    names(filtered_cells) <- translate_sample_ids(names(filtered_cells), sample_ids_map)
+      names(filtered_cells) <- translate_sample_ids(names(filtered_cells), sample_ids_map)
 
-    for (sample_id in unlist(sample_ids_map)) {
-      current_key <- paste0(to_experiment_id, "/", step, "/", sample_id, ".rds")
+      current_key <- paste0(to_experiment_id, "/", step, "/", copy_sample_id, ".rds")
 
+      message("Uploading ", key, "to ", current_key, " with samples translated")
       put_s3_rds(bucket, current_key, aws_config, filtered_cells)
```
Comment (cosa65) [inline pipeline-runner/R/copy-1-copy_s3_objects.R:108]: This change was unnecessary for the fix (filtered cells copy was fine), but it is safer in this shape if we change how filtered-cells works in the future.

The change is to copy individually each file instead of copying the same file for all samples. The previous version worked too because right now we don't really separate the sample files, we just have exact copies filling each of the `step_name/sample_id.rds` files due to I assume a migraiton that wasn't fully finished.

As the TODO in line 70/77 says, once the migraiton is fully done we can actually change this for a way simpler piece of code (we won't even need to download the files).

File: pipeline-runner/R/copy-1-copy_s3_objects.R [hunk 2]
```diff
@@ -48,23 +48,30 @@ copy_cell_sets <- function(from_experiment_id, to_experiment_id, sample_ids_map,
...
+    copy_sample_id <- unlist(sample_ids_map)[[i]]
 
-  s3_copy_by_prefix(
-    pipeline_config$source_bucket,
-    from_experiment_id,
-    pipeline_config$source_bucket,
-    to_experiment_id,
-    pipeline_config$aws_config,
-    key_transform = replace_sample_ids
-  )
```
Comment (cosa65) [inline pipeline-runner/R/copy-1-copy_s3_objects.R:67]: This was the cause of the bug. The reason was that I had misunderstood the shape of the biomage-source files.

I had understood that the sample-specific files in s3 were sample id agnostic inside. This assumption was wrong, there is a `source_data@meta.data$samples` object which is just a vector of the single sample id of the current object. This is probably something we should be able to get rid of in a migration (it is unnecessary) and go back to the original, simpler implementation that just uses s3 copy and is faster instead of this patch that needs to download, modify and upload.

This bug affects any qc runs that are partial (so any runs that don't change the clasiffier's config).

The actual way in which this bug manifested is by showing this message: 
image

Basically, it deleted the filtered cells information because the `source_data@meta.data$samples` was wrong, it was a vector of a repetition of the original experiment's sample id instead of the new one.

As a consequence, in the first step of the new run it assumed no cells had been filtered because that seems to be the default behavior when no filtered-cells object is found. I believe this should probably be changed to error out but don't want to perform more important changes in this PR

File: pipeline-runner/R/handle_data.R [hunk 3]
```diff
@@ -507,9 +507,11 @@ unflatten_cell_sets <- function(cell_sets) {
       cell_set <- list(
         key = children$key[[j]],
         name = children$name[[j]],
-        rootNode = children$rootNode[[j]],
+        rootNode = FALSE,
         color = children$color[[j]],
-        type = children$type[[j]],
+        # Take the type from the parent, some of the child cell sets don't have "type"
+        # Some others do, but all parents do
+        type = cell_sets$type[[i]],
```
Comment (cosa65) [inline pipeline-runner/R/handle_data.R:514]: Some cell sets don't have rootNode and type, some others do. Both of these case scType is fine with.

The problem is that, when copy doesn't find rootNode and type, it sets those values to `NULL`, this case scType is not fine with

File: pipeline-runner/R/copy-1-copy_s3_objects.R [hunk 4]
```diff
@@ -48,23 +48,30 @@ copy_cell_sets <- function(from_experiment_id, to_experiment_id, sample_ids_map,
...
-    for (sample_id in names(sample_ids_map)) {
-      new_key <- gsub(sample_id, sample_ids_map[[sample_id]], new_key)
-    }
+  bucket <- pipeline_config$source_bucket
 
-    return(new_key)
-  }
+  for (i in seq_along(sample_ids_map)) {
+    original_sample_id <- names(sample_ids_map)[[i]]
+    copy_sample_id <- unlist(sample_ids_map)[[i]]
```
Comment (cosa65) [inline pipeline-runner/R/copy-1-copy_s3_objects.R:55]: Oliver: `copy_sample_id <- sample_ids_map[[original_sample_id]]` Is better

File: pipeline-runner/R/copy-1-copy_s3_objects.R [hunk 5]
```diff
@@ -48,23 +48,30 @@ copy_cell_sets <- function(from_experiment_id, to_experiment_id, sample_ids_map,
...
-    new_key <- key
-    for (sample_id in names(sample_ids_map)) {
-      new_key <- gsub(sample_id, sample_ids_map[[sample_id]], new_key)
-    }
+  bucket <- pipeline_config$source_bucket
 
-    return(new_key)
-  }
+  for (i in seq_along(sample_ids_map)) {
+    original_sample_id <- names(sample_ids_map)[[i]]
```
Comment (cosa65) [inline pipeline-runner/R/copy-1-copy_s3_objects.R:54]: `for (original_sample_id in names(sample_ids_map)) {`

File: pipeline-runner/R/copy-1-copy_s3_objects.R [hunk 6]
```diff
@@ -85,16 +92,19 @@ copy_filtered_cells <- function(
   aws_config <- pipeline_config$aws_config
 
   for (step in steps) {
-    # get one of the objects for this experiment, they are all identical
-    key <- paste0(from_experiment_id, "/", step, "/", names(sample_ids_map)[[1]], ".rds")
+    for (i in seq_along(sample_ids_map)) {
+      original_sample_id <- names(sample_ids_map)[[i]]
+      copy_sample_id <- unlist(sample_ids_map)[[i]]
```
Comment (cosa65) [inline pipeline-runner/R/copy-1-copy_s3_objects.R:97]: https://github.com/biomage-org/pipeline/pull/166/files#r1256097356

**Review by ogibson — APPROVED** [review]
